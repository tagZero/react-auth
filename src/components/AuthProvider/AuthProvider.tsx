import React, { createContext, PropsWithChildren, useContext, useEffect, useState } from 'react';
import { AuthProviderPropsType, AuthProviderType } from './AuthProvider.type';

const AuthContext = createContext({} as AuthProviderType);
export const useAuth = () => useContext(AuthContext);

const AuthProvider = ({
  login,
  logout,
  register,
  resetPassword,
  isAuthenticated: overrideIsAuthenticated,
  options: authOptions = {},
  modules: authModules = {},
  messageProvider,
  children
}: PropsWithChildren<AuthProviderPropsType>) => {
  const [token, setToken] = useState<string>();
  const [options, setOptions] = useState({
    ...authOptions,
    authRoute: authOptions?.authRoute || '/auth',
    authenticatedRoute: authOptions?.authenticatedRoute || '/user'
  });
  const [modules, setModules] = useState({
    ...authModules,
    login: {
      path: '/login',
      title: 'Login',
      text: 'Sign in',
      description: 'Please provide your credentials below to login',
      ...authModules?.login
    },
    register: {
      path: '/register',
      title: 'Register',
      text: 'Create an account',
      description: 'Please provide information below to register a new user',
      termsAndConditions: 'I agree to all the',
      termsAndConditionsLink: '/tac',
      termsAndConditionsLinkText: 'terms and conditions',
      successMessage: 'Registration was successful. Please check your email address for activation email.',
      passwordPattern: '(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[!@#$%&*?])[A-Za-z\\d!@#$%&*?]{8,}',
      passwordPatternMessage:
        'Please enter min. 8 characters having at least 1 capital letter, 1 lowercase letter, 1 number and 1 special character (!@#$%&*?)',
      ...authModules?.register
    },
    resetPassword: {
      path: '/reset-password',
      title: 'Reset password',
      text: 'Forgot password?',
      successMessage:
        'Password reset was successful. Please check your email address for password reset email.',
      description: 'Please provide your e-mail address to reset your password',
      ...authModules?.resetPassword
    }
  });

  useEffect(() => {
    if (!document.body.classList.contains('react-auth')) {
      document.body.classList.add('react-auth');
    }
  }, []);

  const isAuthenticated = (): boolean => {
    return !!token;
  };

  const getModulePath = (module: string) => {
    if (options.authRoute === '/') {
      return modules[module].path;
    }
    return options.authRoute + modules[module].path;
  };

  const notify = (notification) => {
    if (messageProvider) {
      return messageProvider(notification);
    } else {
      if (notification.type === 'error') {
        console.error(notification.message);
      } else {
        console.log(notification.message);
      }
    }
  };

  return (
    <AuthContext.Provider
      value={{
        getModulePath,
        isAuthenticated: overrideIsAuthenticated || isAuthenticated,
        login,
        logout,
        modules,
        options,
        register,
        resetPassword,
        setToken,
        notify,
        token
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
