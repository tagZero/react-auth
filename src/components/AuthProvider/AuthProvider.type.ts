export type NotificationType = 'info' | 'warn' | 'error' | 'success';

export interface LoginPropsType {
  email: string;
  password: string;
}

export interface RegisterPropsType extends LoginPropsType {
  firstName: string;
  lastName: string;
  birthDate: number;
}

export interface AuthProviderPropsType {
  isAuthenticated?: () => boolean;
  login?: (props: LoginPropsType, context?: AuthProviderType) => Promise<any>;
  logout?: (context?: AuthProviderType) => Promise<any>;
  register?: (props: RegisterPropsType, context?: AuthProviderType) => Promise<any>;
  resetPassword?: (props: { email: string }, context?: AuthProviderType) => Promise<any>;
  modules?: any;
  options?: any;
}

export interface AuthProviderType extends AuthProviderPropsType {
  authData?: any;
  getModulePath: (module: string) => string;
  notification?: any;
  setAuthData?: (props: any) => any;
  setModules?: (props: any) => any;
  setNotification?: (props: NotificationState) => void;
  setOptions?: (props: any) => any;
  setToken?: (token: string | null) => any;
  token?: string | null;
}

export interface NotificationState {
  type: NotificationType;
  message: string;
}
