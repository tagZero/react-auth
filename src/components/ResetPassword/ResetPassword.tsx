import React from 'react';
import { Link, useHistory } from 'react-router-dom';
import { useAuth } from '../AuthProvider/AuthProvider';

const ResetPassword = () => {
  const history = useHistory();
  const context = useAuth();
  const { resetPassword, modules, getModulePath, options, notify } = context;

  const onSubmit = async (event) => {
    event.preventDefault();
    const email = event.target.email.value;

    try {
      await resetPassword({ email }, context);
      modules.resetPassword.successMessage &&
        notify({ type: 'success', message: modules.resetPassword.successMessage });
      history.push(`${getModulePath('login')}?email=${encodeURIComponent(email)}`);
    } catch (err) {
      notify({ type: 'error', message: err });
    }
  };

  return (
    <main className="auth-container">
      <div className="auth-logo" style={{ backgroundImage: `url(${options.logoUrl}` }} />
      {options.logoTitle ? <div className="auth-title">{options.logoTitle}</div> : null}
      <div className="auth-form-container auth-tile">
        <div className="auth-form-title">{modules.resetPassword.title}</div>
        <p className="auth-form-description">{modules.resetPassword.description}</p>
        <form className="auth-form" onSubmit={onSubmit}>
          <div className="auth-form-row">
            <input type="email" name="email" id="email" placeholder="E-mail" required />
          </div>
          <div className="auth-form-submit">
            <input type="submit" value="Submit" />
          </div>
          <ul className="auth-form-link">
            <li>
              <Link to={getModulePath('login')}>{modules.login.text}</Link>
            </li>
            <li>
              <Link to={getModulePath('register')}>{modules.register.text}</Link>
            </li>
          </ul>
        </form>
      </div>
    </main>
  );
};

export default ResetPassword;
