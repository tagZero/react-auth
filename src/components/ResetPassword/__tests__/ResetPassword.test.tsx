import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import AuthProvider from '../../AuthProvider/AuthProvider';
import ResetPassword from '../ResetPassword';

describe('ResetPassword', () => {
  const tree = () =>
    render(
      <Router>
        <AuthProvider>
          <ResetPassword />
        </AuthProvider>
      </Router>
    );

  test('renders properly', () => {
    const { asFragment } = tree();
    expect(asFragment()).toMatchSnapshot();
  });
});
