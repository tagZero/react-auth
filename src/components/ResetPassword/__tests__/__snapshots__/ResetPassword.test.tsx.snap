// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword renders properly 1`] = `
<DocumentFragment>
  <main
    class="auth-container"
  >
    <div
      class="auth-logo"
    />
    <div
      class="auth-form-container auth-tile"
    >
      <div
        class="auth-form-title"
      >
        Reset password
      </div>
      <p
        class="auth-form-description"
      >
        Please provide your e-mail address to reset your password
      </p>
      <form
        class="auth-form"
      >
        <div
          class="auth-form-row"
        >
          <input
            id="email"
            name="email"
            placeholder="E-mail"
            required=""
            type="email"
          />
        </div>
        <div
          class="auth-form-submit"
        >
          <input
            type="submit"
            value="Submit"
          />
        </div>
        <ul
          class="auth-form-link"
        >
          <li>
            <a
              href="/auth/login"
            >
              Sign in
            </a>
          </li>
          <li>
            <a
              href="/auth/register"
            >
              Create an account
            </a>
          </li>
        </ul>
      </form>
    </div>
  </main>
</DocumentFragment>
`;
